### ISSUE TITLE
GetVisibleBounds on CanvasVirtualImageSource & CanvasVirtualControl

### ISSUE THREAD

**Issue by OvestLabs** [Microsoft/Win2D#574]
This isn't very important, but seems easy to accomplish... I think. 

Could you please expose the `IVirtualSurfaceImageSourceNative.GetVisibleBounds()` method through `CanvasVirtualImageSource` and `CanvasVirtualControl`?

Maybe as a property:

`Rect CanvasVirtualImageSource.VisibleBounds { get; }`
`Rect CanvasVirtualControl.VisibleBounds { get; }`

and even better (no idea if this is possible), expose an event that's raised when visible bounds changes?

**Comment by shawnhar** [comment]
Hi,

This information is exposed through the [CanvasRegionsInvalidatedEventArgs.VisibleRegion](https://microsoft.github.io/Win2D/html/P_Microsoft_Graphics_Canvas_UI_Xaml_CanvasRegionsInvalidatedEventArgs_VisibleRegion.htm) property.  That is a parameter to the RegionsInvalidated event, which is indeed raised whenever the visible bounds change.

**Comment by OvestLabs** [comment]
That's true, but it doesn't report the change when scrolling away from the visible bounds of a `ScrollViewer`. [Here's a repo to demonstrate](https://github.com/OvestLabs/Win2dVisibleBounds) what I'm talking about.

Load it up and start debugging. Pay attention to the output window. When the app loads, you'll see several gray boxes vertically stacked. Scroll down until you see a blue box, which is the `CanvasVirtualControl`. When this box is in the viewport of the `ScrollViewer`, the output window should have the visible region reported by `RegionsInvalidated`. 

Keep scrolling down until the blue box is halfway or completely outside the viewport. Here's the problem: `RegionsInvalidated` is the only source of `VisibleBounds`, but it is not raised when the canvas is outside the visible viewport. It's only raised when portions of the image need to be drawn.

This is why I was asking for `GetVisibleBounds()` exposed in the API. I can calculate the visible region myself, but just thought it would make things cleaner if it was all provided by the API.

**Comment by shawnhar** [comment]
Can you explain more about the scenario where knowing the visible bounds is useful outside the context of wanting to draw something to the control?  If I remember correctly, we put this info in the event args mostly because we couldn't think of a reason to use it anywhere else :-)

Thanks!

**Comment by OvestLabs** [comment]
Imagine you had a `ScrollViewer` with a large scrollable region full of random content, both Xaml and Win2D. There could be several instances of `CanvasVirtualControl` and depending on the source of each control, it may take a while to render a single virtual image. That is, it has thousands of vector geometries. 

When the user scrolls quickly and lands on an area where one of those complex virtual images exist, it will flicker. The area will be blank when scrolled to, and then in a fraction of a second, the image appears.

What I'd like to do is replace the `CanvasVirtualControl` with a low-dpi image of the source when the virtual image is off screen. That way, when scrolled to, the user can get a good idea of what the content is before the clearer high-dpi image loads.

Also, if the user scrolls past these virtual images really quickly (ie, they enter and leave the screen in one swipe), the user can see blurry images instead of blank space. They'll get a good reference point of where they are in scrollable content.

I am computing this on my own. It's no drama. But it would make the code cleaner if the controls I'm building happen to know when they're off the screen.

This is really low priority :)

**Comment by shawnhar** [comment]
Thanks for the extra detail, this makes sense to me now.

One thing to note is that the visible bounds data only reports the current state of what is visible at the present moment, while the invalidated regions has some smarts to predict motion while scrolling and try to render content slightly ahead of when that will actually be needed.  Of course this can still fall behind if rendering is very slow, but this prediction means the invalidated region will typically be larger than (and give better results than) the visible region property.
